src/generator/Generator.ts

```diff
@@ -111,10 +131,55 @@ class Generator {
         }
     }
 
-    private visitBlockStatement(state: VisitorState) {
-        state.body = this.module.block('', state.statements);
+    private visitIfStatement(node: IfStatement, parent: TraversalAncestors, state: VisitorState) {
+        const condition = state.expressionStack.pop();
+
+        if (condition === undefined) {
+            throw new Error('Undefined condition in if statement');
+        }
+
+        const branches = state.branches.get(node);
+
+        if (branches === undefined) {
+            throw new Error('Missing branch statement');
+        }
+
+        const [ifPart, elsePart] = branches;
+        const ifStatement = this.module.if(condition, ifPart, elsePart);
+        const immediateParent = Generator.getImmediateParent(parent);
+
+        if (immediateParent !== undefined && isIfStatement(immediateParent)) {
+            const parentBranches = state.branches.get(immediateParent);
+
+            if (parentBranches === undefined) {
+                throw Error('Parent branches are undefined');
+            }
+
+            parentBranches[1] = ifStatement;
+            state.branches.set(immediateParent, parentBranches);
+        } else {
+            state.statements.push(ifStatement);
+        }
     }
 
+    private visitBlockStatement(state: VisitorState, parent: TraversalAncestors) {
+        const block = this.module.block('', state.statements);
+        state.statements = []; // clear all current statements
+
+        const immediateParent = Generator.getImmediateParent(parent);
```

Comment:
The upcoming code is more or less the exact same as the one in the visitIfStatement method. Extract them to one if possible.

Reply:
I tried that using callbacks but it makes things a bit less readable. The common method would look like this:

```
private isParentIfStatement(parent: TraversalAncestors,
                            success: (immediateParent: IfStatement) => void,
                            failure: () => void) {
    const immediateParent = Generator.getImmediateParent(parent);

    if (immediateParent !== undefined && isIfStatement(immediateParent)) {
        success(immediateParent);
    } else {
        failure();
    }
}
```

And the call from `visitBlockStatement` would then use this new common method:

```
private visitBlockStatement(state: VisitorState, parent: TraversalAncestors) {
    const block = this.module.block('', state.statements);
    state.statements = []; // clear all current statements

    this.isParentIfStatement(parent, (immediateParent) => {
        const parentBranches = state.branches.get(immediateParent);

        if (parentBranches === undefined) {
            state.branches.set(immediateParent, [block, undefined]);
        } else {
            parentBranches[1] = block;
        }
    }, () => {
        state.body = block; // top statement block
    });
}
```
Is this more intention revealing? Or is there a better way to write the common part?

Reply:
I thought more of something like this:
```
private updateImmediateParent(parent: TraversalAncestors,
    onParentBranchUndefined: (state: VisitorState, immediateParent: Node, block: Statement) => void,
    onNonIfStatement: (state: VisitorState, block: Statement) => void,
    state: VisitorState,
    block: Statement) {
    const immediateParent = Generator.getImmediateParent(parent);

    if (immediateParent !== undefined && isIfStatement(immediateParent)) {
        const parentBranches = state.branches.get(immediateParent);

        if (parentBranches === undefined) {
            onParentBranchUndefined(state, immediateParent, block);
        } else {
            parentBranches[1] = block;
        }
    } else {
        onNoIfStatement(state, block);
    }
}
```

The call from ```visitBlockStatement``` would then look like this:
```
private visitBlockStatement(state: VisitorState, parent: TraversalAncestors) {
    const block = this.module.block('', state.statements);
    state.statements = []; // clear all current statements
    updateImmediateParent(parent,
        (state, immediateParent, block) => { state.branches.set(immediateParent, [block, undefined]); },
        (state, block) => { state.body = block; },
        state,
        block);
}
```

The namings in my example aren't the best yet and I'm not yet sure if it is a good idea to implement it that way, but there wouldn't be any more duplicated code.

Reply:
Unfortunately this doesn't work because `visitIfStatement` does more in the else branch than just `parentBranches[1] = block;` So the common method would require another callback. I think we should leave it as it is for the moment.

Reply:
I see, this reduces about four lines of duplicated code. I'm not sure if it's worth the effort because the common function is quite complicated :confused:

Reply:
Fair enough. We probably have to change this code later on anyways. Let's keep it this way for now.